components/script/dom/linkprocessingoptions.rs

```diff
@@ -35,6 +66,155 @@ pub(crate) struct LinkProcessingOptions {
 }
 
 impl LinkProcessingOptions {
+    /// <https://html.spec.whatwg.org/multipage/#extract-links-from-headers>
+    pub(crate) fn extract_links_from_headers(
+        headers: &Option<Serde<HeaderMap>>,
+    ) -> Vec<LinkHeader> {
+        // Step 1. Let links be a new list.
+        let mut links = Vec::new();
+        let Some(headers) = headers else {
+            return links;
+        };
+        // Step 2. Let rawLinkHeaders be the result of getting, decoding, and splitting `Link` from headers.
+        let Some(raw_link_headers) = get_decode_and_split_header_name("Link", headers) else {
+            return links;
+        };
+        // Step 3. For each linkHeader of rawLinkHeaders:
+        for link_header in raw_link_headers {
+            // Step 3.1. Let linkObject be the result of parsing linkHeader. [WEBLINK]
+            let Ok(parsed_link_header) = parse_link_header(&link_header) else {
+                continue;
+            };
+            for link_object in parsed_link_header {
+                let Some(link_object) = link_object else {
+                    // Step 3.2. If linkObject["target_uri"] does not exist, then continue.
+                    continue;
+                };
+                // Step 3.3. Append linkObject to links.
+                links.push(link_object.to_owned());
+            }
+        }
+        // Step 4. Return links.
+        links
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#process-link-headers>
+    pub(crate) fn process_link_headers(link_headers: Vec<LinkHeader>, document: &Document) {
+        // Step 2. For each linkObject in links:
+        for link_object in link_headers {
+            // Step 2.1. Let rel be linkObject["relation_type"].
+            let Some(rel) = link_object.value_for_key_in_link_header("rel") else {
+                continue;
+            };
+            // Step 2.2. Let attribs be linkObject["target_attributes"].
+            //
+            // Not applicable, that's in `link_object.params`
+            // Step 2.3. Let expectedPhase be "media" if either "srcset", "imagesrcset",
+            // or "media" exist in attribs; otherwise "pre-media".
+            // TODO
+            // Step 2.4. If expectedPhase is not phase, then continue.
+            // TODO
+            // Step 2.5. If attribs["media"] exists and attribs["media"] does not match the environment, then continue.
+            if let Some(media) = link_object.value_for_key_in_link_header("media") {
+                if !MediaList::matches_environment(document, media) {
+                    continue;
+                }
+            }
+            // Step 2.6. Let options be a new link processing options with
+            let mut options = LinkProcessingOptions {
```

Comment:
I wonder if it would make sense to only include the spec'd fields and use `..Default::default()`?